tests/transforms/test_add_noise.py

```diff
@@ -0,0 +1,458 @@
+        [tape], _ = add_noise(self.tape_with_prep, self.noise_model_with_prep)
+
+        with qml.queuing.AnnotatedQueue() as q_tape_exp:
+            qml.StatePrep([1, 0], wires=0)
+            qml.ThermalRelaxationError(0.4, 0.4, 0.2, 0.6, wires=0)
+            qml.RZ(0.045, wires=0)
+            qml.RX(0.9, wires=0)
+            qml.RZ(-0.045, wires=0)
+            qml.RY(0.4, wires=1)
+            qml.AmplitudeDamping(0.4, wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RY(0.5, wires=0)
+            qml.AmplitudeDamping(0.4, wires=0)
+            qml.RZ(0.03, wires=1)
+            qml.RX(0.6, wires=1)
+            qml.RZ(-0.03, wires=1)
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+        tape_exp = QuantumScript.from_queue(q_tape_exp)
+
+        assert all(o1.name == o2.name for o1, o2 in zip(tape.operations, tape_exp.operations))
+        assert all(o1.wires == o2.wires for o1, o2 in zip(tape.operations, tape_exp.operations))
+        assert all(
+            np.allclose(o1.parameters, o2.parameters)
+            for o1, o2 in zip(tape.operations, tape_exp.operations)
+        )
+        assert len(tape.measurements) == 1
+        assert (
+            tape.observables[0].name == "Prod"
+            if qml.operation.active_new_opmath()
+            else ["PauliZ", "PauliZ"]
+        )
+        assert tape.observables[0].wires.tolist() == [0, 1]
+        assert tape.measurements[0].return_type is Expectation
+
+
+def test_add_noise_qnode():
+    """Test that a QNode with add_noise decorator gives a different result."""
+    dev = qml.device("default.mixed", wires=2)
+
+    c, n = qml.noise.op_in([qml.RY, qml.RZ]), qml.noise.partial_wires(qml.AmplitudeDamping, 0.4)
+
+    @partial(add_noise, noise_model=qml.NoiseModel({c: n}))
+    @qml.qnode(dev)
+    def f_noisy(w, x, y, z):
+        qml.RX(w, wires=0)
+        qml.RY(x, wires=1)
+        qml.CNOT(wires=[0, 1])
+        qml.RY(y, wires=0)
+        qml.RX(z, wires=1)
+        return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+    @qml.qnode(dev)
+    def f(w, x, y, z):
+        qml.RX(w, wires=0)
+        qml.RY(x, wires=1)
+        qml.CNOT(wires=[0, 1])
+        qml.RY(y, wires=0)
+        qml.RX(z, wires=1)
+        return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+    args = [0.1, 0.2, 0.3, 0.4]
+
+    assert not np.isclose(f_noisy(*args), f(*args))
+
+
+def test_add_noise_dev():
+    """Test if an device transformed by the add_noise function does successfully add noise to
+    subsequent circuit executions"""
+    with qml.queuing.AnnotatedQueue() as q_in_tape:
+        qml.RX(0.9, wires=0)
+        qml.RY(0.4, wires=1)
+        qml.CNOT(wires=[0, 1])
+        qml.RY(0.5, wires=0)
+        qml.RX(0.6, wires=1)
+        qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+        qml.expval(qml.PauliZ(0))
+
+    in_tape = QuantumScript.from_queue(q_in_tape)
+    dev = qml.device("default.qubit", wires=2)
```

Comment:
What happens if the noise isn't supported on the device? Do we perform validation?

Reply:
The device should throw an error during its pre-processing. The dispatch onto a device is dealt with via a transform dispatcher, so it is not straightforward to do any validation on my part.